python/paddle/decomposition/decomp.py

```diff
@@ -279,3 +291,261 @@ def _decompose_subgraph(block, orig_vars, dst_vars, op_filter):
+            return orig_outs
+
+
+def decompose_bwd_op(
+    block, bwd_op, grad_var_to_var_map, fwd_outputs, fwd_inputs
+):
+    '''
+    This API decomposes the bwd_op into a list of primitive ops,
+    and computed the new gradients of the fwd_outputs with respect to the fwd_inputs.
+
+    Args:
+        block (Block): the block to which the bwd_op belongs.
+        bwd_op (pir.Operation): the backward op to be decomposed.
+        grad_var_to_var_map (dict): a dict obtained after distributed processing,
+            which maps the backward grad variable to its corresponding forward variable.
+        fwd_outputs (Value|list(Value)|tuple(Value)): the output Value or Value list/tuple of the forward op.
+        fwd_inputs (Value|list(Value)|tuple(Value)): the input Value or Value list/tuple of the forward op
+
+    Returns:
+        list: a list of Values, the i-th returned Value is the sum of gradients of
+        `fwd_outputs` with respect to the i-th `fwd_inputs`.
+    '''
+
+    if not core._is_bwd_prim_enabled():
+        raise ValueError(
+            "To get composite backward op, please set `core._set_prim_backward_enabled(True)` firstly"
+        )
+    if not isinstance(block, Block):
+        raise TypeError(f"block should be Block, but got type {type(block)}")
+    if not isinstance(bwd_op, pir.Operation):
+        raise TypeError(
+            f"bwd_op should be paddle.pir.Operation, but got type {type(bwd_op)}"
+        )
+    if not isinstance(grad_var_to_var_map, dict):
+        raise TypeError(
+            f"grad_var_to_var_map should be dict which maps grad variable to forward variable, \
+            but got type {type(grad_var_to_var_map)}"
+        )
+    ir_backward.check_type(
+        fwd_outputs,
+        'fwd_outputs',
+        ((pir.Value, pir.OpResult), list, tuple),
+        'paddle.autograd.ir_backward.decompose_bwd_op',
+    )
+    ir_backward.check_type(
+        fwd_inputs,
+        'fwd_outputs',
+        ((pir.Value, pir.OpResult), list, tuple),
+        'paddle.autograd.ir_backward.decompose_bwd_op',
+    )
+
+    fwd_outputs = ir_backward._as_list(fwd_outputs)
+    fwd_inputs = ir_backward._as_list(fwd_inputs)
+
+    # intercept grad_outputs from the original bwd_op
+    # grad_outputs = bwd_op.operands() - fwd_inputs - fwd_outputs
+    grad_outputs = []
+    bwd_inputs = [x.source() for x in bwd_op.operands()]
+    for bwd_input in bwd_inputs:
+        if bwd_input in fwd_inputs or bwd_input in fwd_outputs:
+            continue
+        else:
+            grad_outputs.append(bwd_input)
+
+    # new_fwd_outputs is a subset of fwd_outputs, because some fwd_output does not hold the gradients,
+    # e.g., layer_norm op's output is [out, mean, variance], but only out holds gradient,
+    # therefore, parse the new_fwd_outputs according to grad_outputs and grad_var_to_var_map
+    new_fwd_outputs = []
+    for grad_output in grad_outputs:
+        new_fwd_outputs.append(grad_var_to_var_map[grad_output])
+
```

Comment:
new_fwd_outputs = tuple(grad_var_to_var[grad_output] for grad_output in grad_outputs)